vir/src/gendata.rs

```diff
@@ -117,7 +132,34 @@ pub enum ExprKindGenData<'vir, Curr: 'vir, Next: 'vir> {
 
     Todo(&'vir str),
 }
+
+impl<'vir, Curr, Next> ExprKindGenData<'vir, Curr, Next> {
+    pub fn ty(&self) -> Type<'vir> {
+        match self {
+            ExprKindGenData::Local(l) => l.ty,
+            ExprKindGenData::Field(_, f) => f.ty,
+            ExprKindGenData::Old(e) => e.ty(),
+            ExprKindGenData::Const(c) => c.ty(),
+            ExprKindGenData::Result => &TypeData::Ref, // For now this always seems to be the case
```

Comment:
This seems wrong. `Result` should be the node referring to the value returned by a pure function, which in general is a snapshot.

Reply:
It looks like there are also pure functions returning `Ref`s, for example the function that gets a struct field ref from a struct ref.

I fixed this by changing `Result` to also hold a type; which should allow handling any case.